From 88aba1a5cf384e9ed314729d3a308a2ae92fc521 Mon Sep 17 00:00:00 2001
From: Chao Zeng <chao.zeng@siemens.com>
Date: Thu, 14 Jan 2021 09:21:40 +0800
Subject: [PATCH 9/9] add the gpio alias support

Traditionally the base is given out in first-come-first-serve order.
This might shuffle the numbering of gpios if the probe order changes
So add the gpio alias support in the devicetree to specifies alias ids

Signed-off-by: Chao Zeng <chao.zeng@siemens.com>
---
 drivers/gpio/gpio-davinci.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/gpio/gpio-davinci.c b/drivers/gpio/gpio-davinci.c
index e0b025689625..6b0e72f283e0 100644
--- a/drivers/gpio/gpio-davinci.c
+++ b/drivers/gpio/gpio-davinci.c
@@ -194,6 +194,7 @@ static int davinci_gpio_probe(struct platform_device *pdev)
 	struct davinci_gpio_controller *chips;
 	struct davinci_gpio_platform_data *pdata;
 	struct device *dev = &pdev->dev;
+	int gpio_alias_id;
 
 	pdata = davinci_gpio_get_pdata(pdev);
 	if (!pdata) {
@@ -255,6 +256,16 @@ static int davinci_gpio_probe(struct platform_device *pdev)
 	chips->chip.ngpio = ngpio;
 	chips->chip.base = pdata->no_auto_base ? pdata->base : -1;
 
+/*
+	 * Traditionally the base is given out in first-come-first-serve order.
+	 * This might shuffle the numbering of gpios if the probe order changes.
+	 * So make the base deterministical if the device tree specifies alias
+	 * ids.
+	 */
+	gpio_alias_id = of_alias_get_id(dev->of_node, "gpio");
+	if (gpio_alias_id >= 0)
+		chips->chip.base = gpio_alias_id; 
+
 #ifdef CONFIG_OF_GPIO
 	chips->chip.of_gpio_n_cells = 2;
 	chips->chip.parent = dev;
-- 
2.17.1

